lib/server/services/lhcPeriod/LhcPeriodStatisticsService.js

```diff
@@ -65,6 +65,15 @@ const RUNS_COUNT_SUBQUERY = `
 )
 `;
 
+/*
+ * Subquery to calculate number of data passes associated with given LHC Period
+ */
+const DATA_PASSES_COUNT_SUBQUERY = `
+   (SELECT COUNT(dp.id)
+       FROM data_passes as dp
+       WHERE dp.lhc_period_id = lhcPeriod.id
+   )`;
+
```

Comment:
# Current approach

Sequelize produces this query:
```
SELECT 
    `LhcPeriodStatistics`.`id`,
    `LhcPeriodStatistics`.`avg_center_of_mass_energy` AS `avgCenterOfMassEnergy`, 
    (SELECT GROUP_CONCAT(DISTINCT(r.lhc_beam_energy))
        FROM runs AS r
        WHERE r.lhc_period_id = lhcPeriod.id 
            AND r.definition = 'PHYSICS'
            AND r.run_quality = 'good'
    ) AS `distinctEnergies`, 
    (SELECT r.pdp_beam_type
        FROM runs AS r
        WHERE r.lhc_period_id = lhcPeriod.id 
            AND r.definition = 'PHYSICS'
            AND r.run_quality = 'good'
            AND r.pdp_beam_type IS NOT NULL
        LIMIT 1
    ) AS `beamType`, 
    (SELECT COUNT(dp.id)
        FROM data_passes as dp
        WHERE dp.lhc_period_id = lhcPeriod.id
    ) AS `dataPassesCount`, 
    (SELECT COUNT(r.run_number)
        FROM runs as r
        WHERE r.lhc_period_id = lhcPeriod.id
            AND r.definition = 'PHYSICS'
            AND r.run_quality = 'good'
    ) AS `runsCount`, 
    `lhcPeriod`.`id` AS `lhcPeriod.id`, 
    `lhcPeriod`.`name` AS `lhcPeriod.name`, 
    `lhcPeriod`.`created_at` AS `lhcPeriod.createdAt`, 
    `lhcPeriod`.`updated_at` AS `lhcPeriod.updatedAt` 
FROM 
    `lhc_periods_statistics` AS `LhcPeriodStatistics` 
LEFT OUTER JOIN `lhc_periods` AS `lhcPeriod` 
    ON `LhcPeriodStatistics`.`id` = `lhcPeriod`.`id` 
ORDER BY `lhcPeriod`.`name` DESC;
```

EXPLAIN query output

| id   | select_type        | table                         | type   | possible_keys      | key                | key_len | ref                      | rows  | Extra                           |
|------|--------------------|-------------------------------|--------|--------------------|--------------------|---------|--------------------------|-------|---------------------------------|
|    1 | PRIMARY            | <derived6>                    | ALL    | NULL               | NULL               | NULL    | NULL                     | 19620 | Using temporary; Using filesort |
|    1 | PRIMARY            | lhcPeriod                     | eq_ref | PRIMARY            | PRIMARY            | 4       | LhcPeriodStatistics.id   | 1     |                                 |
|    6 | DERIVED            | p                             | index  | NULL               | PRIMARY            | 4       | NULL                     | 90    | Using index                     |
|    6 | DERIVED            | r                             | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | bookkeeping.p.id         | 218   | Using where                     |
|    6 | DERIVED            | center_of_mass_energy_factors | eq_ref | PRIMARY            | PRIMARY            | 1022    | func                     | 1     | Using where                     |
|    5 | DEPENDENT SUBQUERY | r                             | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | bookkeeping.lhcPeriod.id | 218   | Using where                     |
|    4 | DEPENDENT SUBQUERY | dp                            | ref    | lhc_period_id      | lhc_period_id      | 5       | bookkeeping.lhcPeriod.id | 2     | Using index                     |
|    3 | DEPENDENT SUBQUERY | r                             | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | bookkeeping.lhcPeriod.id | 218   | Using where                     |
|    2 | DEPENDENT SUBQUERY | r                             | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | bookkeeping.lhcPeriod.id | 218   | Using where                     |

# Suggested approach
see code in referenced PR https://github.com/AliceO2Group/Bookkeeping/pull/1368

Sequelize produces this query:
```
SELECT 
    `LhcPeriodStatistics`.`id`, 
    `LhcPeriodStatistics`.`avg_center_of_mass_energy` AS `avgCenterOfMassEnergy`, 
    count(distinct(`dataPass`.`id`)) AS `dataPassesCount`, 
    count(distinct(`run`.`run_number`)) AS `runsCount`, 
    group_concat(distinct(`run`.`lhc_beam_energy`)) AS `distinctEnergies`, 
    group_concat(distinct(`run`.`pdp_beam_type`)) AS `beamType`, 
    `lhcPeriod`.`id` AS `lhcPeriod.id`, 
    `lhcPeriod`.`name` AS `lhcPeriod.name`, 
    `lhcPeriod`.`created_at` AS `lhcPeriod.createdAt`, 
    `lhcPeriod`.`updated_at` AS `lhcPeriod.updatedAt` 
FROM 
    `lhc_periods_statistics` AS `LhcPeriodStatistics` 
LEFT OUTER JOIN `lhc_periods` AS `lhcPeriod` 
    ON `LhcPeriodStatistics`.`id` = `lhcPeriod`.`id`
LEFT OUTER JOIN `data_passes` AS `dataPass` 
    ON `LhcPeriodStatistics`.`id` = `dataPass`.`lhc_period_id` 
LEFT OUTER JOIN `runs` AS `run`
    ON `LhcPeriodStatistics`.`id` = `run`.`lhc_period_id` 
WHERE (
    (`run`.`run_quality` AND `run`.`definition` = 'PHYSICS') 
    OR (`run`.`run_quality` is NULL)) 
GROUP BY `LhcPeriodStatistics`.`id`, `lhcPeriod`.`id` 
ORDER BY `lhcPeriod`.`name` DESC;
```

EXPLAIN query output

| id   | select_type | table                         | type   | possible_keys      | key                | key_len | ref                    | rows  | Extra                           |
|------|-------------|-------------------------------|--------|--------------------|--------------------|---------|------------------------|-------|---------------------------------|
|    1 | PRIMARY     | <derived2>                    | ALL    | NULL               | NULL               | NULL    | NULL                   | 19620 | Using temporary; Using filesort |
|    1 | PRIMARY     | lhcPeriod                     | eq_ref | PRIMARY            | PRIMARY            | 4       | LhcPeriodStatistics.id | 1     |                                 |
|    1 | PRIMARY     | dataPass                      | ref    | lhc_period_id      | lhc_period_id      | 5       | LhcPeriodStatistics.id | 2     | Using index                     |
|    1 | PRIMARY     | run                           | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | LhcPeriodStatistics.id | 218   | Using where                     |
|    2 | DERIVED     | p                             | index  | NULL               | PRIMARY            | 4       | NULL                   | 90    | Using index                     |
|    2 | DERIVED     | r                             | ref    | runs_lhc_period_id | runs_lhc_period_id | 5       | bookkeeping.p.id       | 218   | Using where                     |
|    2 | DERIVED     | center_of_mass_energy_factors | eq_ref | PRIMARY            | PRIMARY            | 1022    | func                   | 1     | Using where                     |

Notes: 
- New approach is actually slightly slower, notes: 
    - adding index on runs.run_quality, runs.defintion DOESN'T help, (in all possible combination)
    - adding index on data_passes.lhc_period_id DOESN'T help
    - adding quais-functional index over virtual column CMPSTR(run_quality, 'good') DOESN'T, (help mariadb doesn't support functional indexes)

Reply:
Thanks for the investigation.
If you have a foreign key as lhc_period_id, it should have a fk and so an index in any case.
One simple question though: why is there LEFT JOIN and not INNER JOIN?

Reply:
`OUTER JOIN` is put by sequelize when the flag `required` (in including association) is omitted or set to false. It's used like this because LHC Periods, which are not associated with `valid` runs, should be kept, in other words, information that particular period has NO good, physics runs associated is still something, besides such a periods, still, might have some Data Passes or Simulation Passes, why... I don't know